// 0g/dist/Assets.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Assets = void 0;
var objectPool_1 = require("./internal/objectPool");
var ResourceHandle_1 = require("./ResourceHandle");
var Assets = /** @class */ (function () {
    function Assets(_loaders) {
        var _this = this;
        this._loaders = _loaders;
        this.handlePool = new objectPool_1.ObjectPool(function () { return new ResourceHandle_1.ResourceHandle(); });
        this.handles = new Map();
        this.load = function (loader, key) {
            var handle = _this.handles.get(_this.getKey(loader, key));
            if (!handle) {
                handle = _this.handlePool.acquire();
                _this.handles.set(_this.getKey(loader, key), handle);
                _this._loaders[loader](key).then(function (value) { return handle.resolve(value); });
            }
            return handle.promise;
        };
        this.immediate = function (loader, key) {
            var handle = _this.handles.get(_this.getKey(loader, key));
            if (!handle)
                return null;
            return handle.value;
        };
        this.getKey = function (loader, key) { return loader + ":" + key; };
    }
    return Assets;
}());
exports.Assets = Assets;
