// 0g/dist/ComponentPool.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ComponentPool = void 0;
var objectPool_1 = require("./internal/objectPool");
var ComponentPool = /** @class */ (function () {
    function ComponentPool(Type, game) {
        var _this = this;
        this.Type = Type;
        this.game = game;
        this.pool = new objectPool_1.ObjectPool(function () { return new _this.Type(); });
    }
    ComponentPool.prototype.acquire = function (initial, id) {
        if (initial === void 0) { initial = {}; }
        var instance = this.pool.acquire();
        this.Type.initialize(instance, initial, id);
        return instance;
    };
    ComponentPool.prototype.release = function (instance) {
        this.pool.release(instance);
    };
    Object.defineProperty(ComponentPool.prototype, "ComponentType", {
        get: function () {
            return this.Type;
        },
        enumerable: false,
        configurable: true
    });
    return ComponentPool;
}());
exports.ComponentPool = ComponentPool;
