// 0g/dist/filters.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isAnyFilter = exports.isChangedFilter = exports.isHasFilter = exports.isNotFilter = exports.isFilter = exports.any = exports.changed = exports.not = exports.has = void 0;
var has = function (Component) { return ({
    Component: Component,
    kind: 'has',
    __isFilter: true,
    toString: function () {
        return "has(" + Component.name + ")";
    },
}); };
exports.has = has;
var not = function (Component) { return ({
    Component: Component,
    kind: 'not',
    __isFilter: true,
    toString: function () {
        return "not(" + Component.name + ")";
    },
}); };
exports.not = not;
var changed = function (Component) { return ({
    Component: Component,
    kind: 'changed',
    __isFilter: true,
    toString: function () {
        return "changed(" + Component.name + ")";
    },
}); };
exports.changed = changed;
var any = function () {
    var Components = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        Components[_i] = arguments[_i];
    }
    return ({
        Components: Components,
        kind: 'any',
        __isFilter: true,
        toString: function () {
            return "any(" + Components.map(function (Comp) { return Comp.name; }).join(', ') + ")";
        },
    });
};
exports.any = any;
var isFilter = function (thing) {
    return thing.__isFilter === true;
};
exports.isFilter = isFilter;
var isNotFilter = function (fil) {
    return fil.kind === 'not';
};
exports.isNotFilter = isNotFilter;
var isHasFilter = function (fil) {
    return fil.kind === 'has';
};
exports.isHasFilter = isHasFilter;
var isChangedFilter = function (fil) { return fil.kind === 'changed'; };
exports.isChangedFilter = isChangedFilter;
var isAnyFilter = function (fil) { return fil.kind === 'any'; };
exports.isAnyFilter = isAnyFilter;
