// 0g/dist/ids.js
"use strict";
/**
 * IDs = 32-bit integers.
 * Lowest 24 bits are the ID itself.
 * Highest 8 bits are the version.
 * An ID with an incremented version is the same ID (same index),
 * but indicates that the resource referred to by an older version
 * is no longer existent.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.incrementIdVersion = exports.setIdVersion = exports.getIdVersion = exports.getIdSignifier = exports.SIGNIFIER_MASK = exports.VERSION_MASK = void 0;
exports.VERSION_MASK = 4278190080;
exports.SIGNIFIER_MASK = 16777215;
/**
 * Gets only the portion of the ID that signifes the
 * resource it represents.
 */
function getIdSignifier(id) {
    // mask out high 8 bits
    return exports.SIGNIFIER_MASK & id;
}
exports.getIdSignifier = getIdSignifier;
/**
 * Gets only the version portion of the ID
 */
function getIdVersion(id) {
    return id >>> 24;
}
exports.getIdVersion = getIdVersion;
function setIdVersion(id, version) {
    version = version % 255 << 24;
    return id | version;
}
exports.setIdVersion = setIdVersion;
function incrementIdVersion(id) {
    return setIdVersion(id, getIdVersion(id) + 1);
}
exports.incrementIdVersion = incrementIdVersion;
