// 0g/dist/Resources.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Resources = void 0;
var objectPool_1 = require("./internal/objectPool");
var logger_1 = require("./logger");
var ResourceHandle_1 = require("./ResourceHandle");
var Resources = /** @class */ (function () {
    function Resources() {
        var _this = this;
        this.handlePool = new objectPool_1.ObjectPool(function () { return new ResourceHandle_1.ResourceHandle(); });
        this.handles = new Map();
        this.getOrCreateGlobalHandle = function (key) {
            var handle = _this.handles.get(key);
            if (!handle) {
                handle = _this.handlePool.acquire();
                _this.handles.set(key, handle);
            }
            return handle;
        };
        this.load = function (key) {
            return _this.getOrCreateGlobalHandle(key).promise;
        };
        this.resolve = function (key, value) {
            _this.getOrCreateGlobalHandle(key).resolve(value);
            logger_1.logger.debug('Resolved resource', key);
        };
        this.immediate = function (key) {
            return _this.getOrCreateGlobalHandle(key).value;
        };
        this.remove = function (key) {
            var value = _this.handles.get(key);
            if (value) {
                _this.handlePool.release(value);
                _this.handles.delete(key);
            }
        };
    }
    return Resources;
}());
exports.Resources = Resources;
