// 0g/dist-esm/input/keyboard.js
var Keyboard = /** @class */ (function () {
    function Keyboard() {
        var _this = this;
        this.keysPressed = new Set();
        this.keysDown = new Set();
        this.keysUp = new Set();
        this.handleKeyDown = function (ev) {
            var key = ev.key;
            // avoid key-repeat triggering?
            if (!_this.keysPressed.has(key)) {
                _this.keysPressed.add(key);
                _this.keysDown.add(key);
            }
        };
        this.handleKeyUp = function (ev) {
            var key = ev.key;
            _this.keysPressed.delete(ev.key);
            _this.keysUp.add(key);
        };
        window.addEventListener('keydown', this.handleKeyDown);
        window.addEventListener('keyup', this.handleKeyUp);
    }
    Keyboard.prototype.getKeyPressed = function (key) {
        return this.keysPressed.has(key);
    };
    Keyboard.prototype.getKeyDown = function (key) {
        return this.keysDown.has(key);
    };
    Keyboard.prototype.getKeyUp = function (key) {
        return this.keysUp.has(key);
    };
    Keyboard.prototype.getAllPressedKeys = function () {
        return this.keysPressed;
    };
    Keyboard.prototype.frame = function () {
        this.keysDown.clear();
        this.keysUp.clear();
    };
    return Keyboard;
}());
export { Keyboard };
export var keyboard = new Keyboard();
